core/src/main/java/org/apache/iceberg/SnapshotProducer.java

```diff
@@ -195,18 +183,26 @@ protected void validate(TableMetadata currentMetadata) {
    * @param metadataToUpdate the base table metadata to apply changes to
    * @return a manifest list for the new snapshot.
    */
-  protected abstract List<ManifestFile> apply(TableMetadata metadataToUpdate);
+  protected abstract List<ManifestFile> apply(TableMetadata metadataToUpdate, Long snapshotId);
```

Comment:
Was thinking apply(TableMetadata metadataToUpdate, Snapshot snapshotId); because we have to do `base.snapshot(snapshotId);` everywhere to get the snapshot to use.

Reply:
Agreed, lets just pass in the entire snapshot.